src/mods/ilbattle.cpp

```diff
@@ -0,0 +1,202 @@
+        mkb::mode_info.stage_time_limit / 2) {  // With time bonus
+        return score + (igt_score + goal_bonus) * 2;
+    } else {
+        return score + (igt_score + goal_bonus);  // Without time bonus
+    }
+}
+
+static u32 convert_battle_length(u8 battle_length_choice) {
+    if (battle_length_choice == 0)
+        return 5 * 60 * 60;
+    else if (battle_length_choice == 1)
+        return 7 * 60 * 60;
+    else if (battle_length_choice == 2)
+        return 10 * 60 * 60;
+    else
+        return 0;
+}
+
+void new_battle() {
+    s_state = IlBattleState::WaitForFirstRetry;
+    battle_frames = 0;
+    best_frames = 0;
+    best_score = 0;
+    buzzer_beater_message = false;
+    buzzer_message_count = 0;
+    battle_length = convert_battle_length(pref::get_il_battle_length());
+}
+
+static void track_first_retry() {
+    bool paused_now = *reinterpret_cast<u32*>(0x805BC474) & 8;
+    if (!paused_now &&
+        (mkb::sub_mode == mkb::SMD_GAME_READY_INIT || mkb::sub_mode == mkb::SMD_GAME_READY_MAIN)) {
+        new_battle();
+        s_state = IlBattleState::BattleRunning;
+    }
+}
+
+static void run_battle_timer() {
+    if (pref::get_il_battle_length() != 3) {  // If the timer isnt endless
+        if (battle_frames < battle_length) {
+            battle_frames++;
+        } else
+            s_state = IlBattleState::BuzzerBeater;
+    } else
+        battle_frames++;  // If the timer is endless
+}
+
+static void track_best() {
+    u32 current_frames = mkb::mode_info.stage_time_frames_remaining;
+    u32 current_score = mkb::balls[mkb::curr_player_idx].score;
+    if (mkb::sub_mode == mkb::SMD_GAME_GOAL_INIT ||
+        mkb::sub_mode == mkb::SMD_GAME_GOAL_MAIN) {            // Goal test
+        if (current_frames > best_frames && !invalid_pause) {  // New best time test, & invalid
+                                                               // pause test
+            best_frames = current_frames;
+        }
+        if (score_calc(current_score) > best_score && !invalid_pause) {  // New best score test, &
+                                                                         // invalid pause test
+            best_score = score_calc(current_score);
+        }
+    }
+}
+
+static void track_invalid_pauses() {
+    bool paused_now = *reinterpret_cast<u32*>(0x805BC474) & 8;
+    if (mkb::mode_info.stage_time_frames_remaining == mkb::mode_info.stage_time_limit - 1) {
+        invalid_pause = false;
+    }
+    if (paused_now) {
+        invalid_pause = true;
+    }
+}
+
+static void track_final_attempt() {
+    bool paused_now = *reinterpret_cast<u32*>(0x805BC474) & 8;
+    // Save time & end battle if: Goal
+    if (mkb::sub_mode == mkb::SMD_GAME_GOAL_INIT || mkb::sub_mode == mkb::SMD_GAME_GOAL_MAIN) {
+        u32 pre_buzzer_time = best_frames;
+        track_best();
+        if (pre_buzzer_time < best_frames) buzzer_beater_message = true;
+        s_state = IlBattleState::BattleDone;
```

Comment:
Could you replace the `buzzer_beater_message` bool with a `BattleDoneBuzzerBeater` state that substitutes `BattleDone` if it's a buzzer beater?